vital/types/local_cartesian.h

```diff
@@ -0,0 +1,182 @@
+   *    cartesian_coordinate[2] : Calculated local cartesian Z coordinate, in meters   (output)
+   */
+  void convert_from_geodetic(vector_3d const& geodetic_coordinate, vector_3d& cartesian_coordinate) const;
+
+  /**
+   * The function convertToGeodetic converts local cartesian
+   * coordinates (X, Y, Z) to geodetic coordinates (latitude, longitude,
+   * and height), according to the WGS84 ellipsoid and local origin parameters.
+   *
+   *    cartesian_coordinate[0] : Local cartesian X coordinate, in meters    (input)
+   *    cartesian_coordinate[1] : Local cartesian Y coordinate, in meters    (input)
+   *    cartesian_coordinate[2] : Local cartesian Z coordinate, in meters    (input)
+   *    geodetic_coordinate[0]  : Calculated longitude value, in degrees     (output)
+   *    geodetic_coordinate[1]  : Calculated latitude value, in degrees      (output)
+   *    geodetic_coordinate[2]  : Calculated height value, in meters         (output)
+   */
+  void convert_to_geodetic(vector_3d const& cartesian_coordinate, vector_3d& geodetic_coordinate) const;
+
+  /**
+   * The function convertFromGeocentric converts geocentric
+   * coordinates according to the WGS84 ellipsoid and local origin parameters.
+   *
+   *    geocentric_coordinate[0] : Geocentric latitude, in meters                       (input)
+   *    geocentric_coordinate[1] : Geocentric longitude, in meters                      (input)
+   *    geocentric_coordinate[2] : Geocentric height, in meters                         (input)
+   *    cartesian_coordinate[0]  : Calculated local cartesian X coordinate, in meters   (output)
+   *    cartesian_coordinate[1]  : Calculated local cartesian Y coordinate, in meters   (output)
+   *    cartesian_coordinate[2]  : Calculated local cartesian Z coordinate, in meters   (output)
+   */
+  void convert_from_geocentric(vector_3d const& geocentric_coordinate, vector_3d& cartesian_coordinate) const;
+
+  /**
+   * The function Convert_Local_Cartesian_To_Geocentric converts local cartesian
+   * coordinates (x, y, z) to geocentric coordinates (X, Y, Z) according to the
+   * current ellipsoid and local origin parameters.
+   *
+   *    cartesian_coordinates[0]  : Local cartesian X coordinate, in meters    (input)
+   *    cartesian_coordinates[1]  : Local cartesian Y coordinate, in meters    (input)
+   *    cartesian_coordinates[2]  : Local cartesian Z coordinate, in meters    (input)
+   *    geocentric_coordinates[0] : Calculated U value, in meters              (output)
+   *    geocentric_coordinates[1] : Calculated v value, in meters              (output)
+   *    geocentric_coordinates[2] : Calculated w value, in meters              (output)
+   */
+  void convert_to_geocentric(vector_3d const& cartesian_coordinate, vector_3d& geocentric_coordinate) const;
+
+private:
+
+  // Ellipsoid Parameters.
+  // All are set to a WGS 84 ellipsoid
+  double semiMajorAxis;
```

Comment:
This is not critical (maybe something for a later branch), but having all of these variables in the header exposes a lot of implementation detail that is not needed here.  We might want to use a PIMPL pattern to move this into the CXX.  That better separates the API from the implementation and prevents everything from needing to be recompiled if someone tweaks one of these internal variables.

Reply:
Yeah, That thought did cross my mind, I just didn't get back to it.
Should be real easy and is a good idea. I will do that

Reply:
This may not be an issue in this case, but using the PIMPL pattern can limit the ability to specialize the class by subclassing since all the really interesting methods are not accessible from the base class.

Reply:
Good Tip, I am comfortable encapsulating this code into a pimpl, its well encapsulated and has not needed to change in over a decade. May want to make the class accessable at some point to users but it should not need tweeked if someone wants to extend the local_cartersian class